tests/unit/test_loaders.py

```diff
@@ -206,9 +294,116 @@ def test_load_service_model_uses_provided_type_name(self):
+        # sanity check that one of the paths is a ZipPath
+        # since it can't be checked directly
+        self.assertTrue(
+            any(isinstance(path, ZipPath) for path in loader.search_paths)
+        )
+
+    @pytest.mark.skipif(
+        sys.version_info < (3, 9), reason="Python version < 3.9"
+    )
+    def test_zipped_load_data(self):
+        path = os.path.join(os.path.dirname(__file__), 'data', 'Archive.zip')
+        loader = Loader(
+            extra_search_paths=[path], include_default_search_paths=False
+        )
+        data = loader.load_data('foo')
+        self.assertEqual(len(data), 3)
+        self.assertTrue('test_key_1' in data)
+
+    @pytest.mark.skipif(
+        sys.version_info < (3, 9), reason="Python version < 3.9"
+    )
+    def test_zipped_data_not_found_raises_exception(self):
+        path = os.path.join(os.path.dirname(__file__), 'data', 'Archive.zip')
+        loader = Loader(
+            extra_search_paths=[path], include_default_search_paths=False
+        )
+        with self.assertRaises(DataNotFoundError):
+            loader.load_data('cheese')
+
+    @pytest.mark.skipif(
+        sys.version_info < (3, 9), reason="Python version < 3.9"
+    )
+    def test_zipped_model_not_found_raises_exception(self):
+        path = os.path.join(os.path.dirname(__file__), 'data', 'Archive.zip')
+        loader = Loader(
+            extra_search_paths=[path], include_default_search_paths=False
+        )
+        with self.assertRaises(DataNotFoundError):
+            loader.load_service_model('bestServiceEver', 'service-2')
+
+    @pytest.mark.skipif(
+        sys.version_info < (3, 9), reason="Python version < 3.9"
+    )
+    def test_load_zipped_service_model(self):
+        path = os.path.join(
+            os.path.dirname(__file__), 'data', 'Archive.zip', 'foo'
+        )
+        loader = Loader(
+            extra_search_paths=[path], include_default_search_paths=False
+        )
+        model = loader.load_service_model('myCoolService', 'service-2')
+        self.assertIn("foo", model)
+
+    @pytest.mark.skipif(
+        sys.version_info < (3, 9), reason="Python version < 3.9"
+    )
+    def test_zipped_list_available_services(self):
+        path = os.path.join(
+            os.path.dirname(__file__), 'data', 'Archive.zip', 'foo'
+        )
+        loader = Loader(
+            extra_search_paths=[path], include_default_search_paths=False
+        )
+        available_services = loader.list_available_services('service-2')
+        self.assertEqual(
+            available_services, ['dynamodb', 'ec2', 'myCoolService']
+        )
+
+    @pytest.mark.skipif(
+        sys.version_info < (3, 9), reason="Python version < 3.9"
+    )
+    def test_zipped_list_api_versions(self):
+        path = os.path.join(
+            os.path.dirname(__file__), 'data', 'Archive.zip', 'foo'
+        )
+        loader = Loader(
+            extra_search_paths=[path], include_default_search_paths=False
+        )
```

Comment:
There's a few repeating patterns here. When we find that in test cases, it's usually a good sign to move it into a single function or fixture. That will keep the codebase slimmer, make future changes easier to organize, and reduce the risk of accidental bugs being missed.

Reply:
sure. i came from a job previously where a lot of folks felt the opposite. that hard coding tests with repeatable information made it easier to read, debug and catch bugs because of all the hand writing required. I don't necessarily feel strongly either way, but I see the merits of both sides. Regardless, if its preferred to make tests as dry as possible, I'm happy to do so.